### ISSUE TITLE
Showing command hint on error is much harder in v2

### ISSUE THREAD

**Issue by xealot** [alecthomas/kingpin#34]
In version 2 some of the base functionality has been un-exported. Instead of showing the entire usage when a command fails I wrote a function to show just the subcommand help if we could understand the command that far. This was the code:

```c
func generateHint(app *kingpin.Application, args []string) (string, error) {
	output := bytes.NewBuffer(nil)
	context := kingpin.Tokenize(args)

	token := &kingpin.Token{0, ""}
	for token.Type != kingpin.TokenEOL {
		token = context.Peek()
		if token.Type == kingpin.TokenArg {
			if _, ok := commands[token.String()]; ok {
				app.CommandUsage(output, token.String())
				return output.String(), nil
			} else {
				// If we hit a real arg and it isn't a sub-command exit.
				break
			}
		} else {
			// Skip one additional time to pass over Flag value
			context.Next()
		}
		context.Next()
	}

	return "", fmt.Errorf("No hints")
}
```

I don't know if this is the right way to do this, but that is no longer possible. Since our command has about 12 subcommands it's very irritating to see the entire help tree when really they're struggling with a single command.

**Comment by alecthomas** [comment]
Hi there.

Believe it or not, this will get much simpler in v2, but it's not quite ready yet. v2 splits parsing and execution into multiple passes, building an [intermediate structure](https://github.com/alecthomas/kingpin/blob/v2-unstable/parser.go#L51) that contains the result of the parse. This contains the exact flags, commands and arguments that the parser sees, so you will not need to use heuristics to figure out what the `args[]string` contains.

However, there are some things holding up progress:

1. I'm trying to figure out a better way to display help. As you mention, dumping the full list of commands is a bit overwhelming. My current thought is to just dump each level of the tree as each command is encountered.
2. I want to expose the internal command tree structure publicly, and use text/template to generate the help from that structure. This would allow users to override the help in any way they like. This looks like it will require a fair bit of work though.

So, basically, it's not ready yet, but it should hopefully be much better than v1.

**Comment by xealot** [comment]
Having tried what must be dozens of these types of libs having a text template is something I've always wanted. It seems like a very natural approach and I'm somewhat surprised it's not more common.

I agree that if you can descend into the subcommands and show relevant help it would be much more intuitive.

**Comment by xealot** [comment]
:+1: NOICE!
